source/fuzz/fuzzer_pass_donate_modules.cpp

```diff
@@ -1007,7 +1009,96 @@ void FuzzerPassDonateModules::PrepareInstructionForDonation(
       input_operands));
 }
 
-void FuzzerPassDonateModules::AddLivesafeFunction(
+bool FuzzerPassDonateModules::CreateLoopLimiterInfo(
+    opt::IRContext* donor_ir_context, const opt::BasicBlock& loop_header,
+    const std::map<uint32_t, uint32_t>& original_id_to_donated_id,
+    protobufs::LoopLimiterInfo* out) {
+  assert(loop_header.IsLoopHeader() && "|loop_header| is not a loop header");
+
+  protobufs::LoopLimiterInfo loop_limiter;
+
+  // Grab the loop header's id, mapped to its donated value.
+  loop_limiter.set_loop_header_id(
+      original_id_to_donated_id.at(loop_header.id()));
+  // Get fresh ids that will be used to load the loop limiter, increment
+  // it, compare it with the loop limit, and an id for a new block that
+  // will contain the loop's original terminator.
+  loop_limiter.set_load_id(GetFuzzerContext()->GetFreshId());
+  loop_limiter.set_increment_id(GetFuzzerContext()->GetFreshId());
+  loop_limiter.set_compare_id(GetFuzzerContext()->GetFreshId());
+  loop_limiter.set_logical_op_id(GetFuzzerContext()->GetFreshId());
+
+  // We are creating a branch from the back-edge block to the merge block. Thus,
+  // if merge block has any OpPhi instructions, we might need to adjust
+  // them.
+  //
+  // Note that the loop might have an unreachable back-edge block. This means
+  // that the loop can't iterate, so we don't need to adjust anything.
+  if (const auto back_edge_block_id =
+          TransformationAddFunction::GetBackEdgeBlockId(donor_ir_context,
+                                                        loop_header.id())) {
```

Comment:
Would it be too ugly if we did an early return true here (setting `out` as well), instead of a large conditional?

Reply:
Fixed in c89d2e6.